packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -446,5 +452,685 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
   }
 }
 
+// Renders the content of the selection menu and maintains the page state.
+class _CupertinoTextSelectionToolbarContent extends StatefulWidget {
+  const _CupertinoTextSelectionToolbarContent({
+    Key key,
+    @required this.children,
+    @required this.isArrowPointingDown,
+  }) : assert(children != null),
+       // This ignore is used because .isNotEmpty isn't compatible with const.
+       assert(children.length > 0), // ignore: prefer_is_empty
+       super(key: key);
+
+  final List<Widget> children;
+  final bool isArrowPointingDown;
+
+  @override
+  _CupertinoTextSelectionToolbarContentState createState() => _CupertinoTextSelectionToolbarContentState();
+}
+
+class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSelectionToolbarContent> with TickerProviderStateMixin {
+  // Controls the fading of the buttons within the menu during page transitions.
+  AnimationController _controller;
+  int _page = 0;
+  int _nextPage;
+
+  void _handleNextPage() {
+    _controller.reverse();
+    _controller.addStatusListener(_statusListener);
+    _nextPage = _page + 1;
+  }
+
+  void _handlePreviousPage() {
+    _controller.reverse();
+    _controller.addStatusListener(_statusListener);
+    _nextPage = _page - 1;
+  }
+
+  void _statusListener(AnimationStatus status) {
+    if (status != AnimationStatus.dismissed) {
+      return;
+    }
+
+    setState(() {
+      _page = _nextPage;
+      _nextPage = null;
+    });
+    _controller.forward();
+    _controller.removeStatusListener(_statusListener);
+  }
+
+  @override
+  void didUpdateWidget(_CupertinoTextSelectionToolbarContent oldWidget) {
+    // If the children are changing, the current page should be reset.
+    if (widget.children != oldWidget.children) {
+      _page = 0;
```

Comment:
Maybe also reset _nextPage to null? (I don't think this is a visual bug, but it may be strange during debugging if this is set to something that's no longer valid.)